website/docs/src/content/patterns/tools.mdx

```diff
@@ -0,0 +1,102 @@
+---
+title: Tools
+description: Using tools with GenSX
+---
+
+# Using tools with GenSX
+
+Using tools provides a powerful and flexible way to extend the capabilities of an LLM. GenSX offers several different ways for you to work with tools:
+
+- Using the `GSXChatCompletion` component from [@gensx/openai](../component-reference/openai.mdx) or [@gensx/anthropic](../component-reference/anthropic.mdx) for fully automated tool calling
+- Using the `OpenAIChatCompletion` or `AnthropicChatCompletion` components from [@gensx/openai](../component-reference/openai.mdx) or [@gensx/anthropic](../component-reference/anthropic.mdx) and following the same pattern as described in the [OpenAI](https://platform.openai.com/docs/guides/function-calling?api-mode=responses) and [Anthropic](https://docs.anthropic.com/en/docs/build-with-claude/tool-use/overview) docs
+- Using tools with the [Vercel AI SDK](https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling) via the [@gensx/vercel-ai-sdk](../component-reference/vercel-ai.mdx) package
+
+This guide will focus on the first option: using `GSXTool` and the `GSXChatCompletion` component for automated tool calling. The examples below show how to work with the [@gensx/openai](../component-reference/openai.mdx) package but you can follow a similar pattern with the [@gensx/anthropic](../component-reference/anthropic.mdx) package.
+
+## Defining a tool
+
+Start by defining the tool. A `GSXTool` contains four main pieces:
+
+- `name`: The name of the tool
+- `description`: A description of the tool
+- `schema`: The schema of the tool, defined as a Zod object
+- `run`: The function that will be called when the tool is invoked by an LLM
+
+Here's an example of a basic weather tool with mocked functionality:
+
+```tsx
+import { GSXTool } from "@gensx/openai";
+import { z } from "zod";
+
+// Define the schema as a Zod object
+const weatherSchema = z.object({
+  location: z.string(),
+});
+
+// Use z.infer to get the type for our parameters
+type WeatherParams = z.infer<typeof weatherSchema>;
+
+// Create the tool with the correct type - using the schema type, not the inferred type
+const weatherTool = new GSXTool({
+  name: "get_weather",
+  description: "get the weather for a given location",
+  schema: weatherSchema,
+  run: async ({ location }: WeatherParams) => {
+    console.log("getting weather for", location);
+    const weather = ["sunny", "cloudy", "rainy", "snowy"];
+    return Promise.resolve({
+      weather: weather[Math.floor(Math.random() * weather.length)],
+    });
+  },
+});
```

Comment:
I tihnk its worth mentioning that this can just be the params too:

```suggestion
const weatherTool: gensx.GSXToolParams = {
  name: "get_weather",
  description: "get the weather for a given location",
  schema: weatherSchema,
  run: async ({ location }: WeatherParams) => {
    console.log("getting weather for", location);
    const weather = ["sunny", "cloudy", "rainy", "snowy"];
    return Promise.resolve({
      weather: weather[Math.floor(Math.random() * weather.length)],
    });
  },
}
```

Reply:
I think I missed that we changed this to ToolParams at some point. Should it be ToolProps, to be more consistent with naming conventions for the rest of the programming model ?

Reply:
Here's a PR for the rename assuming we decide to move forward with it: https://github.com/gensx-inc/gensx/pull/469

Reply:
also, added a note in the doc regarding Jeremy's original point